### PR TITLE
NIAD-797 First end-to-end test

### PR DESCRIPTION
**Pull request by mattd-kainos**

### CODE REVIEW

File: e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Hashtable;
+
+public class MessageQueue {
+
+    public static void sendToMhsInboundQueue(String messageContent) throws Exception {
+        String broker = System.getenv().getOrDefault("GP2GP_AMQP_BROKERS", "amqp://localhost:5672");
+        String queueName = System.getenv().getOrDefault("GP2GP_MHS_INBOUND_QUEUE", "inbound");
+
+        Hashtable<String, String> hashtable = new Hashtable<>();
+        hashtable.put("connectionfactory.SBCF", broker);
```
Comment (Broom94) [inline e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java:20]: Will this work with AWS?

File: e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        hashtable.put("queue.QUEUE", queueName);
+        hashtable.put(Context.INITIAL_CONTEXT_FACTORY, "org.apache.qpid.jms.jndi.JmsInitialContextFactory");
+        Context context = new InitialContext(hashtable);
+
+        ConnectionFactory cf = (ConnectionFactory) context.lookup("SBCF");
```
Comment (Broom94) [inline e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java:25]: We could rename "SBCF" to "CF" or ConnectionFactory?

**Review by Broom94 — APPROVED** [review]

File: e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java (hunk 2 above)
Comment (mattd-kainos) [reply e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java:25]: Yes, "SBCF" is Service Bus Connection Factory - I ripped this off of some Azure documentation :D

**Review by michael-kainos — DISMISSED** [review]

File: service/build.gradle [hunk 3]
```diff
@@ -20,6 +22,10 @@ lombok {
 	config['lombok.log.fieldName'] = 'LOGGER'
 }
 
+springBoot {
+	buildInfo()
```
Comment (mattd-kainos) [inline service/build.gradle:26]: remove, for testing a new ticket

File: service/build.gradle [hunk 4]
```diff
@@ -54,15 +60,23 @@ dependencies {
...
 	testImplementation "org.assertj:assertj-core:3.18.1"
 	testImplementation "org.testcontainers:testcontainers:1.15.0"
-	testImplementation 'org.springframework:spring-jms:5.2.6.RELEASE'
-	testImplementation 'org.apache.qpid:qpid-jms-client:0.51.0'
 	testImplementation 'commons-io:commons-io:2.6'
 }
 
+tasks.withType(Test) {
+	testLogging {
+		events "passed", "skipped", "failed"
```
Comment (mattd-kainos) [inline service/build.gradle:68]: remove, already in elliot's branch

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusRepositoryTest.java [hunk 5]
```diff
@@ -33,7 +30,7 @@
     @Test
     public void When_AddingNewEhrExtractStatus_Expect_EhrExtractStatusRetrievableByIdFromDatabase() {
         Instant now = Instant.now();
-        ehrExtractStatusRepository.save(new EhrExtractStatus(EhrStatusConstants.EXTRACT_ID,
+        ehrExtractStatusRepository.save(new EhrExtractStatus(
```
Comment (mattd-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusRepositoryTest.java:33]: Add a test the unique index is defined correctly

Comment (mattd-kainos) [reply service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusRepositoryTest.java:33]: Unit test that mongo error is handled for index error

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/UnsupportedInteractionException.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package uk.nhs.adaptors.gp2gp.mhs;
+
+public class UnsupportedInteractionException extends RuntimeException {
+
+    public UnsupportedInteractionException(String interactionId) {
+        super("Unsupported interaction id " + interactionId);
+    }
+
+    public static class InvalidInboundMessageException extends RuntimeException {
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/UnsupportedInteractionException.java:9]: Refactor / autogen mistake? Should be separate classes

**Review by petervdm — APPROVED** [review]

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusRepositoryTest.java [hunk 7]
```diff
@@ -17,24 +15,29 @@
...
 import org.springframework.test.context.junit.jupiter.SpringExtension;
+import uk.nhs.adaptors.gp2gp.common.mongo.MongoClientConfiguration;
 
 @ExtendWith({SpringExtension.class, MongoDBExtension.class, ActiveMQExtension.class})
 @SpringBootTest
 @DirtiesContext
 public class EhrExtractStatusRepositoryTest {
+
+    @Autowired
+    private MongoClientConfiguration mongoClientConfiguration;
```
Comment (michael-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusRepositoryTest.java:26]: remove unused variable

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/common/amqp/MessageQueueTest.java [hunk 8]
```diff
@@ -33,14 +10,22 @@
...
+import static org.mockito.Mockito.verify;
 
 @SpringBootTest
 @ExtendWith({MongoDBExtension.class, ActiveMQExtension.class})
 @DirtiesContext
 public class MessageQueueTest {
-    private static final String EHR_EXTRACT_REQUEST_TEST_FILE = "/ehrExtractRequest.json";
     private static final String SOAP_HEADER = "<soap:Header></soap:Header>";
     private static final long TIMEOUT = 5000L;
     private static final long EXPECTED_COUNT = 1;
```
Comment (michael-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/common/amqp/MessageQueueTest.java:31]: Remove unused variable

File: e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import javax.naming.InitialContext;
+import java.util.Hashtable;
+
+public class MessageQueue {
+
+    public static void sendToMhsInboundQueue(String messageContent) throws Exception {
+        String broker = System.getenv().getOrDefault("GP2GP_AMQP_BROKERS", "amqp://localhost:5672");
+        String queueName = System.getenv().getOrDefault("GP2GP_MHS_INBOUND_QUEUE", "inbound");
+
+        Hashtable<String, String> hashtable = new Hashtable<>();
```
Comment (michael-kainos) [inline e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java:19]: Consider replacing this Hashtable with the newer java.util.Map

Comment (mattd-kainos) [reply e2e-tests/src/main/java/uk/nhs/adaptors/gp2gp/MessageQueue.java:19]: InitialContext requires a Hashtable

**Review by Broom94 — DISMISSED** [review]

**Review by petervdm — APPROVED** [review]

**Review by Broom94 — APPROVED** [review]

**Pull request merged**